processing/src/main/java/io/druid/segment/DimensionIndexer.java

```diff
@@ -0,0 +1,350 @@
+ * the second ordering will be referred to as "Sorted".
+ *
+ * The unsorted ordering is used during ingestion, within the IncrementalIndex's TimeAndDims keys; the encodings
+ * are built as rows are ingested, taking the order in which new dimension values are seen.
+ *
+ * The generation of a sorted encoding takes place during segment creation when indexes are merged/persisted.
+ * The sorted ordering will be used for dimension value arrays in that context and when reading from
+ * persisted segments.
+ *
+ * Note that after calling the methods below that deal with sorted encodings,
+ * - getSortedEncodedValueFromUnsorted()
+ * - getUnsortedEncodedValueFromSorted()
+ * - getSortedIndexedValues()
+ * - convertUnsortedEncodedArrayToSortedEncodedArray()
+ *
+ * calling processRowValsToUnsortedEncodedArray() afterwards can invalidate previously read sorted encoding values
+ * (i.e., new values could be added that are inserted between existing values in the ordering).
+ *
+ *
+ * Thread Safety
+ * --------------------
+ * Each DimensionIndexer exists within the context of a single IncrementalIndex. Before IndexMerger.persist() is
+ * called on an IncrementalIndex, any associated DimensionIndexers should be thread-safe.
+ *
+ * In current usage, this means that an implementation should allow multiple threads to add data to the indexer via
+ * processRowValsToUnsortedEncodedArray() and allow multiple threads to read data via methods that only deal with
+ * unsorted encodings.
+ *
+ * As mentioned in the "Sorting and Ordering" section, writes and calls to the sorted encoding
+ * methods should not be interleaved: the sorted encoding methods should only be called when it is known that
+ * writes to the indexer will no longer occur.
+ *
+ * Index merge and persist are currently handled by a single thread.
+ *
+ * After IndexMerger.persist() is called on an IncrementalIndex, any indexers associated with that IncrementalIndex
+ * are no longer required to be thread-safe.
+ *
+ * The sorted encoding methods are not used outside of index merging/persisting (single-threaded context, and
+ * no new events will be added to the indexer).
+ *
+ * @param <EncodedType> class of the encoded values
+ * @param <ActualType> class of the actual values
+ *
+ */
+public interface DimensionIndexer<EncodedType extends Comparable<EncodedType>, EncodedTypeArray, ActualType extends Comparable<ActualType>>
+{
+  /**
+   * Given a single row value or list of row values (for multi-valued dimensions), update any internal data structures
+   * with the ingested values and return the row values as an array to be used within a TimeAndDims key.
+   *
+   * For example, the dictionary-encoded String-type column will return an int[] containing a dictionary ID.
+   *
+   * The value within the returned array should be encoded if applicable, i.e. as instances of EncodedType.
+   *
+   * NOTE: This function can change the internal state of the DimensionIndexer.
```

Comment:
I think there should be a bit more highlighted and add a link to proper usage (as explained in the comments above)